xtask/src/themelint.rs

```diff
@@ -0,0 +1,194 @@
+        if self.fg.is_some() {
+            self.fg.unwrap()
+        } else if self.bg.is_some() {
+            self.bg.unwrap()
+        } else {
+            "LINTER_ERROR_NO_RULE"
+        }
+    }
+
+    fn check_difference(
+        theme: &Theme,
+        a: &'static str,
+        b: &'static str,
+        messages: &mut Vec<String>,
+    ) {
+        let theme_a = theme.get(a);
+        let theme_b = theme.get(b);
+        if theme_a == theme_b {
+            messages.push(format!("$THEME: `{}` and `{}` cannot be equal", a, b));
+        }
+    }
+
+    fn check_existence(rule: &Rule, theme: &Theme, messages: &mut Vec<String>) {
+        let found_fg = rule.found_fg(theme);
+        let found_bg = rule.found_bg(theme);
+
+        if !rule.check_both && (found_fg || found_bg) {
+            return;
+        }
+        if !found_fg || !found_bg {
+            let mut missing = vec![];
+            if !found_fg {
+                missing.push("`fg`");
+            }
+            if !found_bg {
+                missing.push("`bg`");
+            }
+            let entry = if !rule.check_both && !found_fg && !found_bg {
+                missing.join(" or ")
+            } else {
+                missing.join(" and ")
+            };
+            messages.push(format!(
+                "$THEME: missing {} for `{}`",
+                entry,
+                rule.rule_name()
+            ))
+        }
+    }
+}
+
+pub fn lint(file: String) -> Result<(), DynError> {
+    if file.contains("base16") {
+        println!("Skipping base16: {}", file);
+        return Ok(());
+    }
+    let path = path::themes().join(file.clone() + ".toml");
+    let theme = std::fs::read(&path).unwrap();
+    let theme: Theme = toml::from_slice(&theme).expect("Failed to parse theme");
+
+    let mut messages: Vec<String> = vec![];
+    get_rules().iter().for_each(|lint| match lint {
+        Require::Existence(rule) => Rule::check_existence(rule, &theme, &mut messages),
+        Require::Difference(a, b) => Rule::check_difference(&theme, a, b, &mut messages),
+    });
+
+    if !messages.is_empty() {
+        messages.iter().for_each(|m| {
+            let theme = file.clone();
+            let message = m.replace("$THEME", theme.as_str());
+            println!("{}", message);
+        });
+        Err(messages.len().to_string().into())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn lint_all() -> Result<(), DynError> {
+    let files = Loader::read_names(path::themes().as_path());
+    let mut errors = vec![];
+    let files_count = files.len();
+    files.into_iter().for_each(|path| {
+        if let Err(err) = lint(path.replace(".toml", "")) {
+            let errs: i32 = err.to_string().parse().expect("Errors must be integral");
+            errors.push(errs)
+        }
+    });
+    println!("{} of {} themes had issues", errors.len(), files_count);
+    if !errors.is_empty() {
+        Err(errors.iter().sum::<i32>().to_string().into())
```

Comment:
This seemed pretty much useless to me such that it returns some numbers which may confuse users.
```
...
varua: missing `fg` or `bg` for `ui.virtual.ruler`
41 of 60 themes had issues
Error: "95"
```

I have no idea what `95` means when I am looking at the output, if anything `41 of 60 themes had issues` or `lint issues found in theme` may be a better error, in any case I prefer there isn't the `Error: ` line but it's fine to have one as long as it is somewhat useful.

Also, the count doesn't seemed necessary as well in any place, I don't recall seeing rustc or similar linters count how many lint errors are there.

Reply:
The printout of error is because xtask main is or Result return type I think? I totally agree that the total count printout makes little sense. I've moved the `x of y themes had issues` printout to be the message of the error.